http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -951,4 +881,74 @@ object LeonardoService {
         labels.combineAll
       case None => Right(params)
     }
+
+  private[service] def augmentClusterRequest(serviceAccountInfo: ServiceAccountInfo,
+                                             googleProject: GoogleProject,
+                                             clusterName: ClusterName,
+                                             userEmail: WorkbenchEmail,
+                                             clusterRequest: ClusterRequest,
+                                             clusterImages: Set[ClusterImage]): ClusterRequest = {
+    val userJupyterExt = clusterRequest.jupyterExtensionUri match {
+      case Some(ext) => Map("notebookExtension" -> ext.toUri)
+      case None      => Map.empty[String, String]
+    }
+
+    // add the userJupyterExt to the nbExtensions
+    val updatedUserJupyterExtensionConfig = clusterRequest.userJupyterExtensionConfig match {
+      case Some(config) => config.copy(nbExtensions = config.nbExtensions ++ userJupyterExt)
+      case None         => UserJupyterExtensionConfig(userJupyterExt, Map.empty, Map.empty, Map.empty)
+    }
+
+    // transform Some(empty, empty, empty, empty) to None
+    val updatedClusterRequest = clusterRequest.copy(
+      userJupyterExtensionConfig =
+        if (updatedUserJupyterExtensionConfig.asLabels.isEmpty)
+          None
+        else
+          Some(updatedUserJupyterExtensionConfig)
+    )
+
+    addClusterLabels(serviceAccountInfo, googleProject, clusterName, userEmail, updatedClusterRequest, clusterImages)
+  }
+
+  private[service] def calculateAutopauseThreshold(autopause: Option[Boolean], autopauseThreshold: Option[Int], autoFreezeConfig: AutoFreezeConfig): Int =
+    autopause match {
+      case None =>
+        autoFreezeConfig.autoFreezeAfter.toMinutes.toInt
+      case Some(false) =>
+        autoPauseOffValue
+      case _ =>
+        if (autopauseThreshold.isEmpty) autoFreezeConfig.autoFreezeAfter.toMinutes.toInt
+        else Math.max(autoPauseOffValue, autopauseThreshold.get)
+    }
+
+  private[service] def addClusterLabels(serviceAccountInfo: ServiceAccountInfo,
+                                        googleProject: GoogleProject,
+                                        clusterName: ClusterName,
+                                        creator: WorkbenchEmail,
+                                        clusterRequest: ClusterRequest,
+                                        clusterImages: Set[ClusterImage]): ClusterRequest = {
+    // create a LabelMap of default labels
+    val defaultLabels = DefaultLabels(
+      clusterName,
+      googleProject,
+      creator,
+      serviceAccountInfo.clusterServiceAccount,
+      serviceAccountInfo.notebookServiceAccount,
+      clusterRequest.jupyterUserScriptUri,
+      clusterRequest.jupyterStartUserScriptUri,
+      clusterImages.map(_.imageType).filterNot(_ == Welder).headOption
+    ).toJson.asJsObject.fields.mapValues(labelValue => labelValue.convertTo[String])
```

Comment:
Should this use `DefaultLabels.toMap`?

Reply:
I didn't want to replace existing code too much. But I guess this one should be pretty safe to do